pkg/xds/experimental/client/client.go

```diff
@@ -0,0 +1,603 @@
+					TypeUrl:                obsReq.typeUrl,
+					ResourceNamesSubscribe: obsReq.resourceNames,
+				}
+				log.Debug("Delta send", "req", req)
+				err = delta.Send(req)
+			}
+			if err != nil {
+				log.Error("Stream/Delta send", "error", err)
+				errCh <- err
+				backoff.Wait(ctx)
+			}
+		case resp, ok := <-c.responseQueue:
+			if !ok {
+				return
+			}
+			log.Debug("Stream receive", "resp", resp)
+			upsertedResources, err := c.handleResponse(resp)
+			if err != nil {
+				log.Error("Failed to handle response", "error", err)
+				errCh <- err
+				err := c.sendNACK(ctx, stream, resp, err)
+				if err != nil {
+					log.Error("Failed to send NACK", "error", err)
+				}
+				backoff.Wait(ctx)
+				continue
+			}
+			err = c.upsertAndDeleteMissing(upsertedResources)
+			if err != nil {
+				log.Error("Failed to update resources", "error", err)
+				continue
+			}
+			err = c.sendACK(ctx, stream, resp, upsertedResources)
+			if err != nil {
+				log.Error("Failed to send ACK", "error", err)
+				continue
+			}
+		case resp, ok := <-c.deltaResponseQueue:
+			if !ok {
+				return
+			}
+			log.Debug("Delta receive", "resp", resp)
+			upsertedResources, err := c.handleDeltaResponse(resp)
+			if err != nil {
+				log.Error("Failed to handle response", "error", err)
+				errCh <- err
+				err := c.sendDeltaNACK(ctx, delta, resp, err)
+				if err != nil {
+					log.Error("Failed to send NACK", "error", err)
+				}
+				backoff.Wait(ctx)
+				continue
+			}
+			typeUrl := resp.GetTypeUrl()
+			respDelRes := resp.GetRemovedResourceNames()
+			deletedResources := make([]string, 0, len(respDelRes))
+			for _, res := range respDelRes {
+				deletedResources = append(deletedResources, res.GetName())
+			}
+			c.log.Debug("cache TX", "typeUrl", typeUrl, "upserted", upsertedResources, "deleted", deletedResources)
+			ver, updated, _ := c.cache.TX(typeUrl, upsertedResources, deletedResources)
+			c.log.Debug("cache TX", "typeUrl", typeUrl, "ver", ver, "updated", updated)
+			err = c.sendDeltaACK(ctx, delta, resp)
+			if err != nil {
+				log.Error("Failed to send ACK", "error", err)
+				continue
+			}
+		}
+	}
+}
+
+// nameToResource maps a resource name to a proto representation of the resource.
+type nameToResource map[string]proto.Message
+
+func (c *XDSClient) handleDeltaResponse(resp *discoverypb.DeltaDiscoveryResponse) (nameToResource, error) {
+	var errs error
+	ret := make(nameToResource, len(resp.GetResources()))
+	for _, res := range resp.GetResources() {
+		name := res.GetName()
+		msg, _, err := parseResource(resp.GetTypeUrl(), res.GetResource())
+		if err != nil {
+			errs = errors.Join(errs, err)
+			continue
+		}
+		ret[name] = msg
+	}
+	return ret, errs
+}
+
+// handleResponse decodes the resources based on TypeUrl and updates the cache.
+func (c *XDSClient) handleResponse(resp *discoverypb.DiscoveryResponse) (map[string]nameToResource, error) {
+	var errs error
+	upsertedResources := make(map[string]nameToResource, 8)
```

Comment:
How we determined the initial length of 8?

Reply:
Simplified the code to support one type per message (as defined in the specification)